10x Martian CLI wrapper and mkfastq WDL

mckinsel opened the pull request:
This PR includes a wrapper around the 10x martian workflow executor that lets you call individual stages from the command line with open source code.

It also has a WDL that implements the `cellranger mkfastq` command using this wrapper.

mckinsel commented on 10x-wdl/wdls/make_fastq.wdl:
```diff
@@ -0,0 +1,567 @@
+task make_fastqs_preflight_local {
+  File run_path
+  Boolean check_executables
+  Array[Int]? lanes
```
Yes, the lanes argument is optional.

mckinsel commented on 10x-wdl/docker-10x/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+
+COPY bcl2fastq /usr/bin/bcl2fastq
+
+ADD cellranger-2.0.0.tar.gz /opt/tenx
```
Agree, but the cellranger repo is missing some things that are present in the cellranger tarball. Also, we'd have to compile everything in the cellranger repo, and my understanding is that that's pretty difficult.

I'd prefer to at least download the tarball from 10x rather than copying it, but it seems like the download site is generating a fresh access token for you each time you load the page: https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest

dshiga replied:
Gotcha. I thought we weren't actually using any compiled cellranger code though, only python and bash. What's missing from the cellranger repo that we need? Let's not wait for it for this PR but do you think it's worth asking them to add it?

mckinsel replied:
What we need to know is how to correctly compile something like this: https://github.com/10XGenomics/cellranger/blob/master/tenkit/lib/go/src/q30count/main.go

I think their build system is complex; they want the cellranger tarball to just work in lots of environments, so I assume there are lots of tricks.

We are using compiled components of cellranger, but only those for which we have the source.

mckinsel commented on 10x-wdl/docker-10x/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+COPY run_in_10x_env.bash /opt/tenx/run_in_10x_env.bash
+
+
+RUN adduser --home /home/tenx_runner tenx_runner 
```
The user isn't needed. I've taken it out.

mckinsel commented on 10x-wdl/docker-10x/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+RUN /bin/bash -c "source /opt/tenx/cellranger-2.0.0/sourceme.bash && pip install martian_cli"
+
+WORKDIR /home/tenx_runner
+ENTRYPOINT ["/opt/tenx/run_in_10x_env.bash"]
```
I removed the entrypoint. The wdl tasks already call the script directly, so removing the entrypoint shouldn't change anything.

mckinsel commented on 10x-wdl/wdls/make_fastq.wdl:
```diff
@@ -0,0 +1,567 @@
+  # /cromwell-execution/etc/etc path we get from File inputs don't do that. So hard
+  # link the path to $CWD/fastq_path before we start working with it.
+  command {
+    cp -al '${fastq_path}' fastq_path
```
I think generally no. The cellranger code stores paths as strings in some of its output files, and all this copying to $PWD/fastq_path is happening so those paths are valid across different tasks. If I use a symbolic link, cellranger dereferences it and stores a path like "/cromwell-execution/uuid/this-stage/inputs/........".

mckinsel commented on 10x-wdl/wdls/make_fastq.wdl:
```diff
@@ -0,0 +1,567 @@
+  }
+
+  runtime {
+    docker: "10x_hca_wdl:latest"
```
Should this really go in the HCA docker hub account?

mckinsel commented on 10x-wdl/wdls/make_fastq.wdl:
```diff
@@ -0,0 +1,567 @@
+      --bcl2fastq_version '${bcl2fastq_version}' \
+      --split_file '${split_file}'
+    
+    jq -r '.[]' merged_file_paths > merged_file_paths.lines
```
Added a comment.

mckinsel commented on 10x-wdl/wdls/make_fastq.wdl:
```diff
@@ -0,0 +1,567 @@
+      jq -c '.[] | map(sub(".*(?=fastq_path\/)"; "")) ' | \
+      split -l 1 -a 2 - merged_file_paths_output
+    
+    while read fastq; do
```
Added a comment.

mckinsel commented on 10x-wdl/wdls/make_fastq.wdl:
```diff
@@ -0,0 +1,567 @@
+
+  command {
+    mkdir fastq_path
+    for fastq in ${sep=' ' input_files}; do
```
Added a comment.

mckinsel commented on 10x-wdl/wdls/make_fastq.wdl:
```diff
@@ -0,0 +1,567 @@
+    samples="${dollar}(jq -s '.[].sample' ${sep=' ' split_files} | jq -rs 'join(" ")')"
+    lanes="${dollar}(jq -s '.[].lane | tostring' ${sep=' ' split_files} | jq -rs 'join(" ")')"
+
+    echo '[${sep=',' qc_jsons}]' | \
```
Added a comment.

mckinsel commented on 10x-wdl/docker-10x/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+
+USER root
+
+COPY bcl2fastq /usr/bin/bcl2fastq
```
Yeah it's just the illumina bcl2fastq. I replaced this with a step that downloads and installs it from Illumina.

kbergin commented:
@dshiga @mckinsel Just your friendly PO popping in to say that adding documentation for the pipeline is in this ticket https://elastc.com/c/5URiWaJK/38-10x-control-pipeline-design-doc and isn't technically in the sprint, but we could pull it in if it doesn't make sense to merge in the code without it. :)

kbergin commented:
I've made a separate ticket for point  # 1 mentioned in Dave's comments above about separating out the MRO parsing that's done on the fly. It is here: https://elastc.com/c/DeUf0rub/172-stop-parsing-mro-on-the-fly-and-write-out-python

mckinsel commented:
Hey @dshiga,

Thanks for the comments! Here are responses to your list:

1. This is a good idea but a fair bit of work, so we're moving it to a separate ticket.
2. Added a couple tests for the parser.
3. Added some READMEs throughout, including for the WDLs.
4. Also added some description of the package.
5. Moved everything to the 10x directory
6. I think I don't know how to do this after the PR ha already been made...
7. Added an example input for the make fastq WDL.

kbergin commented:
@mckinsel I think re: number 6 you can just edit the top comment on the PR webpage.

dshiga commented:
Hey @mckinsel, thanks for making those changes. The extra documentation is really helpful and it's great to see those tests. Could you add a script to run the tests in docker, similar to this one 
https://github.com/HumanCellAtlas/secondary-analysis/blob/master/test/test.sh
Then you could remove the .mro files and just use the ones in the docker.

dshiga commented:
Oh, about (6), I see that you actually did put a description when you opened the PR - it shows up as a comment. I just didn't notice it earlier. So never mind.
